File: mobile_client/lib/models/game/player_answer.dart
import 'package:mobile_client/constants/enum/question_type.dart';

abstract class PlayerAnswer {
  final QuestionType questionType;
  PlayerAnswer(this.questionType);

  Map<String, dynamic> toJson();
}

class MultiChoicePlayerAnswer extends PlayerAnswer {
  final List<bool> data;

  MultiChoicePlayerAnswer({required this.data})
      : super(QuestionType.multiChoice);

  @override
  Map<String, dynamic> toJson() {
    return {
      'questionType': questionType.toJson(),
      'data': data,
    };
  }
}

class OpenEndedPlayerAnswer extends PlayerAnswer {
  final String data;

  OpenEndedPlayerAnswer({required this.data}) : super(QuestionType.openEnded);

  @override
  Map<String, dynamic> toJson() {
    return {
      'questionType': questionType.toJson(),
      'data': data,
    };
  }
}

class EstimationPlayerAnswer extends PlayerAnswer {
  final int data;

  EstimationPlayerAnswer({required this.data}) : super(QuestionType.estimation);

  @override
  Map<String, dynamic> toJson() {
    return {
      'questionType': questionType.toJson(),
      'data': data,
    };
  }
}

class DrawingPlayerAnswer extends PlayerAnswer {
  final String data;

  DrawingPlayerAnswer({required this.data}) : super(QuestionType.drawing);

  @override
  Map<String, dynamic> toJson() {
    return {
      'questionType': questionType.toJson(),
      'data': data,
    };
  }
}

File: common/interfaces/player-answer.ts
import { QuestionType } from '../enum/question-type';

interface MultiChoicePlayerAnswer {
    questionType: QuestionType.MultiChoice;
    data: boolean[];
}

interface OpenEndedPlayerAnswer {
    questionType: QuestionType.OpenEnded;
    data: string;
}

interface EstimationPlayerAnswer {
    questionType: QuestionType.Estimation;
    data: number;
}

interface DrawingPlayerAnswer {
    questionType: QuestionType.Drawing;
    data: string;
}

export type PlayerAnswer = MultiChoicePlayerAnswer | OpenEndedPlayerAnswer | EstimationPlayerAnswer | DrawingPlayerAnswer;
